src/main/scala/akka/persistence/cassandra/ConfigSessionProvider.scala

```diff
@@ -0,0 +1,155 @@
+    .setNewConnectionThreshold(
+      HostDistance.REMOTE,
+      connectionPoolConfig.getInt("new-connection-threshold-remote"))
+    .setMaxRequestsPerConnection(
+      HostDistance.LOCAL,
+      connectionPoolConfig.getInt("max-requests-per-connection-local"))
+    .setMaxRequestsPerConnection(
+      HostDistance.REMOTE,
+      connectionPoolConfig.getInt("max-requests-per-connection-remote"))
+    .setConnectionsPerHost(
+      HostDistance.LOCAL,
+      connectionPoolConfig.getInt("connections-per-host-core-local"),
+      connectionPoolConfig.getInt("connections-per-host-max-local"))
+    .setConnectionsPerHost(
+      HostDistance.REMOTE,
+      connectionPoolConfig.getInt("connections-per-host-core-remote"),
+      connectionPoolConfig.getInt("connections-per-host-max-remote"))
+    .setPoolTimeoutMillis(
+      connectionPoolConfig.getInt("pool-timeout-millis"))
+
+  def clusterBuilder(clusterId: String)(implicit ec: ExecutionContext): Future[Cluster.Builder] = {
+    lookupContactPoints(clusterId).map { cp =>
+      val b = Cluster.builder
+        .addContactPointsWithPorts(cp.asJava)
+        .withPoolingOptions(poolingOptions)
+        .withQueryOptions(new QueryOptions().setFetchSize(fetchSize))
+      protocolVersion match {
+        case None    => b
+        case Some(v) => b.withProtocolVersion(v)
+      }
+
+      val username = config.getString("authentication.username")
+      if (username != "") {
+        b.withCredentials(
+          username,
+          config.getString("authentication.password"))
+      }
+
+      val localDatacenter = config.getString("local-datacenter")
+      if (localDatacenter != "") {
+        b.withLoadBalancingPolicy(
+          new TokenAwarePolicy(
+            DCAwareRoundRobinPolicy.builder.withLocalDc(localDatacenter).build()))
+      }
+
+      val truststorePath = config.getString("ssl.truststore.path")
+      if (truststorePath != "") {
+        val trustStore = StorePathPasswordConfig(
+          truststorePath,
+          config.getString("ssl.truststore.password"))
+
+        val keystorePath = config.getString("ssl.keystore.path")
+        val keyStore: Option[StorePathPasswordConfig] =
+          if (keystorePath != "") {
+            val keyStore = StorePathPasswordConfig(
+              keystorePath,
+              config.getString("ssl.keystore.password"))
+            Some(keyStore)
+          } else None
+
+        val context = SSLSetup.constructContext(trustStore, keyStore)
+
+        b.withSSL(JdkSSLOptions.builder.withSSLContext(context).build())
+      }
+
+      b
+    }
+  }
+
+  /**
+   * Subclass may override this method to perform lookup the contact points
+   * of the Cassandra cluster asynchronously instead of reading them from the
+   * configuration.
+   *
+   * @param clusterId the configured `cluster-id` to lookup
+   */
+  def lookupContactPoints(clusterId: String)(implicit ec: ExecutionContext): Future[immutable.Seq[InetSocketAddress]] = {
```

Comment:
@huntc you would create a subclass of `ConfigSessionProvider` and override this `lookupContactPoints` method